js/src/snapshot.ts

```diff
@@ -0,0 +1,453 @@
+      fields: [
+        ['fee_credit', 'u64'],
+        ['fee_address', 'u256'],
+        ['stake_seeds', SeedRange],
+        ['unstake_seeds', SeedRange],
+        ['stake_accounts_balance', 'u64'],
+        ['unstake_accounts_balance', 'u64'],
+        ['active', 'u8'],
+      ],
+    },
+  ],
+  [
+    ValidatorPubKeyAndEntry,
+    {
+      kind: 'struct',
+      fields: [
+        ['pubkey', 'u256'],
+        ['entry', Validator],
+      ],
+    },
+  ],
+  [
+    MaintainerPubKeyAndEntry,
+    {
+      kind: 'struct',
+      fields: [
+        ['pubkey', 'u256'],
+        ['entry', [0]],
+      ],
+    },
+  ],
+  [
+    RewardDistribution,
+    {
+      kind: 'struct',
+      fields: [
+        ['treasury_fee', 'u32'],
+        ['validation_fee', 'u32'],
+        ['developer_fee', 'u32'],
+        ['st_sol_appreciation', 'u32'],
+      ],
+    },
+  ],
+  [
+    FeeRecipients,
+    {
+      kind: 'struct',
+      fields: [
+        ['treasury_account', 'u256'],
+        ['developer_account', 'u256'],
+      ],
+    },
+  ],
+  [
+    Validators,
+    {
+      kind: 'struct',
+      fields: [
+        ['entries', [ValidatorPubKeyAndEntry]],
+        ['maximum_entries', 'u32'],
+      ],
+    },
+  ],
+  [
+    Maintainers,
+    {
+      kind: 'struct',
+      fields: [
+        ['entries', [MaintainerPubKeyAndEntry]],
+        ['maximum_entries', 'u32'],
+      ],
+    },
+  ],
+  [
+    Solido,
+    {
+      kind: 'struct',
+      fields: [
+        ['lido_version', 'u8'],
+
+        ['manager', 'u256'],
+
+        ['st_sol_mint', 'u256'],
+
+        ['exchange_rate', ExchangeRate],
+
+        ['sol_reserve_authority_bump_seed', 'u8'],
+        ['stake_authority_bump_seed', 'u8'],
+        ['mint_authority_bump_seed', 'u8'],
+        ['rewards_withdraw_authority_bump_seed', 'u8'],
+
+        ['reward_distribution', RewardDistribution],
+
+        ['fee_recipients', FeeRecipients],
+
+        ['metrics', Metrics],
+
+        ['validators', Validators],
+
+        ['maintainers', Maintainers],
+      ],
+    },
+  ],
+]);
+
+export const getSolido = (solidoInstanceDataBuffer: Buffer) => {
+  const deserialized = deserializeUnchecked(
+    schema,
+    Solido,
+    solidoInstanceDataBuffer
+  );
+
+  return deserialized;
+};
+
+export const getSnapshot = async (
+  connection: Connection,
+  solidoInstanceAccountInfo: AccountInfo<Buffer>,
+  reserveAccountInfo: AccountInfo<Buffer>,
+  programAddresses: ProgramAddresses
+): Promise<Snapshot> => {
+  const solido = getSolido(solidoInstanceAccountInfo.data);
+
+  const reserveAccountRentExemptionBalance =
+    await connection.getMinimumBalanceForRentExemption(
+      reserveAccountInfo.data.byteLength
+    );
+
+  const reserveAccountBalance: Lamports = {
+    lamports: new BN(
+      reserveAccountInfo.lamports - reserveAccountRentExemptionBalance
+    ),
+  };
+
+  const {
+    value: { amount },
+  } = await connection.getTokenSupply(programAddresses.stSolMintAddress);
```

Comment:
We can’t use the `getTokenSupply` if we want to avoid torn reads. It might happen that we read the Solido account, and we find that the stake accounts together hold 100 SOL (and 100 stSOL is minted for that, but that is not stored in the Solido instance). Then somebody withdraws, now there is only 50 SOL left and 50 stSOL minted. Then we call `getTokenSupply`, and we find 50 stSOL. Now the snapshot claims to have 100 SOL but only 50 stSOL. But that’s not true, and even worse, there never even was a time at which it _was_ true!

To avoid this, we have to use `getMultipleAccounts`. That’s the only tool the Solana RPC gives us to avoid torn reads. We can include the stSOL mint in the accounts to query, and then use `encoding: 'jsonParsed`. It returns data of this shape for SPL token mints:

```json
"value": [
  {
    "data": {
      "parsed": {
        "info": {
          "decimals": 9,
          "freezeAuthority": null,
          "isInitialized": true,
          "mintAuthority": "8kRRsKezwXS21beVDcAoTmih1XbyFnEAMXXiGXz6J3Jz",
          "supply": "1367470251242903"
        },
        "type": "mint"
      },
      "program": "spl-token",
      "space": 82
    },
    "executable": false,
    "lamports": 1461600,
    "owner": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
    "rentEpoch": 268
  }
]
```

(If you want to be really strict, we also need to avoid `connection.getMinimumBalanceForRentExemption`, because it performs a separate read, and read and parse `SYSVAR_RENT` instead. But the rent is unlikely to change, so I can live with that being a separate read. I expect it would be faster to do everything in a single `getMultipleAccounts` read though, it saves a network round trip.)